nativelib/src/main/resources/scala-native/gc/commix/Synchronizer.c

```diff
@@ -0,0 +1,208 @@
+    switch (ex->ExceptionRecord->ExceptionCode) {
+    case EXCEPTION_ACCESS_VIOLATION:
+        ULONG_PTR addr = ex->ExceptionRecord->ExceptionInformation[1];
+        if (SafepointInstance == (void *)addr) {
+            Synchronizer_wait();
+            return EXCEPTION_CONTINUE_EXECUTION;
+        }
+        fprintf(stderr, "Cought exception code %p in GC exception handler\n",
+                (void *)(uintptr_t)ex->ExceptionRecord->ExceptionCode);
+        fflush(stdout);
+        StackTrace_PrintStackTrace(ex);
+    // pass-through
+    default:
+        return EXCEPTION_CONTINUE_SEARCH;
+    }
+}
+#else
+#ifdef __APPLE__
+#define SAFEPOINT_TRAP_SIGNAL SIGBUS
+#else
+#define SAFEPOINT_TRAP_SIGNAL SIGSEGV
+#endif
+#define THREAD_WAKUP_SIGNAL SIGCONT
+static struct sigaction defaultAction;
+static sigset_t threadWakupSignals;
+static void SafepointTrapHandler(int signal, siginfo_t *siginfo, void *uap) {
+    if (siginfo->si_addr == SafepointInstance) {
+        Synchronizer_wait();
+    } else {
+        fprintf(stderr,
+                "Unexpected signal %d when accessing memory at address %p\n",
+                signal, siginfo->si_addr);
+        StackTrace_PrintStackTrace();
+        defaultAction.sa_handler(signal);
+    }
+}
+#endif
+
+static void SetupPageFaultHandler() {
+#ifdef _WIN32
+    // Call it as last exception handler
+    AddVectoredExceptionHandler(1, &SafepointTrapHandler);
+#else
+    sigemptyset(&threadWakupSignals);
+    sigaddset(&threadWakupSignals, THREAD_WAKUP_SIGNAL);
+    sigprocmask(SIG_BLOCK, &threadWakupSignals, NULL);
+    assert(sigismember(&threadWakupSignals, THREAD_WAKUP_SIGNAL));
+
+    struct sigaction sa;
+    memset(&sa, 0, sizeof(struct sigaction));
+    sigemptyset(&sa.sa_mask);
+    sa.sa_sigaction = &SafepointTrapHandler;
+    sa.sa_flags = SA_SIGINFO | SA_RESTART;
+    if (sigaction(SAFEPOINT_TRAP_SIGNAL, &sa, &defaultAction) == -1) {
+        perror("Error: cannot setup safepoint synchronization handler");
+        exit(errno);
+    }
+#endif
+}
+
+static void Synchronizer_SuspendThread(MutatorThread *thread) {
+    assert(thread == currentMutatorThread);
+#ifdef _WIN32
+    if (!ResetEvent(thread->wakeupEvent)) {
+        fprintf(stderr, "Failed to reset event %lu\n", GetLastError());
+    }
+    if (WAIT_OBJECT_0 != WaitForSingleObject(thread->wakeupEvent, INFINITE)) {
+        fprintf(stderr, "Error: suspend thread");
+        exit(GetLastError());
+    }
+#else
+    int signum;
+    if (0 != sigwait(&threadWakupSignals, &signum)) {
+        perror("Error: sig wait");
+        exit(errno);
+    }
+    assert(signum == THREAD_WAKUP_SIGNAL);
+#endif
+}
+
+static void Synchronizer_WakupThread(MutatorThread *thread) {
+#ifdef _WIN32
+    assert(thread != currentMutatorThread);
+    if (!SetEvent(thread->wakeupEvent)) {
+        fprintf(stderr, "Failed to set event %lu\n", GetLastError());
+    }
+#else
+    int status = pthread_kill(thread->thread, THREAD_WAKUP_SIGNAL);
```

Comment:
```suggestion
    int status = pthread_kill(thread->thread, THREAD_WAKEUP_SIGNAL);
```
This maybe in Immix too.